spacy_llm/tasks/ner.py

```diff
@@ -203,10 +253,108 @@ def scorer(
     ) -> Dict[str, Any]:
         return get_ner_prf(examples)
 
+    @property
+    def _Example(self) -> type[SpanExample]:
+        return SpanExample
+
     def _create_prompt_example(self, example: Example) -> SpanExample:
         """Create an NER prompt example from a spaCy example."""
         entities = defaultdict(list)
         for ent in example.reference.ents:
             entities[ent.label_].append(ent.text)
 
         return SpanExample(text=example.reference.text, entities=entities)
+
+
+class COTNERTask(SpanTask[COTSpanExample]):
+    def __init__(
+        self,
+        labels: List[str],
+        template: str,
+        description: Optional[str] = None,
+        prompt_examples: Optional[List[COTSpanExample]] = None,
+        label_definitions: Optional[Dict[str, str]] = None,
+        normalizer: Optional[Callable[[str], str]] = None,
+        alignment_mode: Literal["strict", "contract", "expand"] = "contract",
+        case_sensitive_matching: bool = False,
+        single_match: bool = False,
+    ):
+        super().__init__(
+            labels=labels,
+            template=template,
+            description=description,
+            label_definitions=label_definitions,
+            prompt_examples=prompt_examples,
+            normalizer=normalizer,
+            alignment_mode=alignment_mode,
+            case_sensitive_matching=case_sensitive_matching,
+            single_match=single_match,
+        )
+
+    def initialize(
+        self,
+        get_examples: Callable[[], Iterable["Example"]],
+        nlp: Language,
+        labels: List[str] = [],
+        **kwargs: Any,
+    ) -> None:
+        """Initialize the NER task, by auto-discovering labels.
+
+        Labels can be set through, by order of precedence:
+
+        - the `[initialize]` section of the pipeline configuration
+        - the `labels` argument supplied to the task factory
+        - the labels found in the examples
+
+        get_examples (Callable[[], Iterable["Example"]]): Callable that provides examples
+            for initialization.
+        nlp (Language): Language instance.
+        labels (List[str]): Optional list of labels.
+        """
+
+        examples = get_examples()
+
+        if not labels:
+            labels = list(self._label_dict.values())
+
+        if not labels:
+            label_set = set()
+
+            for eg in examples:
+                for ent in eg.reference.ents:
+                    label_set.add(ent.label_)
+            labels = list(label_set)
+
+        self._label_dict = {self._normalizer(label): label for label in labels}
+
+    def _format_response(self, response: str) -> Iterable[Tuple[str, Iterable[str]]]:
+        """Parse raw string response into a structured format"""
+        output: dict[str, list[str]] = defaultdict(list)
+        assert self._normalizer is not None
+        for line in response.strip().split("\n"):
+            entity = SpanReason.from_str(line)
+            if entity:
```

Comment:
1. I'd rename `entity` to `span_reason` or similar.
2. `SpanReason.from_str()` only returns a `SpanReason` instance if `line` was parsed successfully for an entity, right?

Reply:
Yeah, it does